crates/parcel_plugin_rpc/src/nodejs/napi.rs

```diff
@@ -0,0 +1,49 @@
+use std::sync::mpsc::Receiver;
+use std::sync::mpsc::Sender;
+use std::thread;
+
+use napi::threadsafe_function::ThreadsafeFunction;
+use napi::threadsafe_function::ThreadsafeFunctionCallMode;
+use napi::Env;
+use napi::JsUnknown;
+use napi::Status;
+use serde::de::DeserializeOwned;
+
+// Generic method to create a "resolve" javascript function to
+// return the value from the thread safe function
+pub fn create_callback<Returns: DeserializeOwned + 'static>(
+  env: &Env,
+  reply: Sender<Returns>,
+) -> napi::Result<JsUnknown> {
+  let callback = env
+    .create_function_from_closure("callback", move |ctx| {
+      let response = ctx
+        .env
+        .from_js_value::<Returns, JsUnknown>(ctx.get::<JsUnknown>(0)?)?;
+
+      if reply.send(response).is_err() {
+        return Err(napi::Error::from_reason("Unable to send rpc response"));
+      }
+
+      ctx.env.get_undefined()
+    })?
+    .into_unknown();
+
+  Ok(callback)
+}
+
+pub fn forward_threadsafe_function_to_channel<T: Send>(
```

Comment:
This does the opposite of what the name says

Reply:
Are you sure we need this, vs. just calling the function?

Reply:
Might not be needed, will play around with it

Reply:
If the threadsafe function is called directly from the `RpcConnection` it runs on the bundler thread and blocks until JS returns a result. This effectively serializes the requests to the workers.

To solve this I have 1 rust thread for 1 js worker thread where I am calling the threadsafe function within its own dedicated thread. This ensures the request/response of the tsfn does not block the caller thread.

I guess this could be avoided if we use async Rust with tokio - but it's not really necessary to involve an async runtime yet

Reply:
I think it makes a lot of use-cases a lot easier to just call the function, because in a lot of cases you'll call the method and want to wait for the response. If you need the thread to be unblocked then there can be an async point at the call site, but most of the time I think you'd want to wait.

Reply:
Even if we use an async runtime and make the call site `async`, the threadsafe function will block the Tokio worker (because it isn't async). 

```rust
impl RpcConnectionNodejs {
  pub async fn ping() {
    self.threadsafe_function.call(()).unwrap() // Will block the Tokio worker
  }
}
```

Wrapping it in a channel + dedicated thread allows us to use async channels and allow the call site to not be blocking.

```rust
impl RpcConnectionNodejs {
  pub async fn ping() {
    let (tx, mut rx) = tokio::oneshot();
    self.tx_threadsafe_function.send(RpcMessage::Ping{ response: tx });
    rx.await
  }
}
```

Reply:
Right but you can just spawn if you don't want to block

Reply:
The other benefit is the channels buffer messages that will be sent to the worker prior to the worker being initialized. This allows for the workers to be initialized lazily.